File: internal/dto/provider/thread_config.go
package provider

// ThreadConfigPatch 是 thread 配置的局部更新载荷，指针字段表示"不传则不覆盖"。
type ThreadConfigPatch struct {
	Model       *string `json:"model,omitempty"`
	Effort      *string `json:"effort,omitempty"`
	Personality *string `json:"personality,omitempty"`
	Approvals   *string `json:"approvals,omitempty"`
}

// ThreadConfigValues 是 thread 配置的值快照，用于展示当前生效或覆盖配置。
type ThreadConfigValues struct {
	Model     string `json:"model,omitempty"`
	Effort    string `json:"effort,omitempty"`
	Approvals string `json:"approvals,omitempty"`
}

// ThreadConfig 是 thread 完整配置视图，包含覆盖值和最终生效值。
type ThreadConfig struct {
	ThreadID               string             `json:"threadId"`
	Provider               string             `json:"provider,omitempty"`
	SupportsThreadOverride bool               `json:"supportsThreadOverride"` // 当前 provider 是否支持 thread 级配置覆盖。
	AvailableModels        []string           `json:"availableModels"`        // 当前 provider session 实时返回的全部可选模型。
	Override               ThreadConfigValues `json:"override"`               // 用户显式设置的覆盖配置。
	Effective              ThreadConfigValues `json:"effective"`              // 合并后最终生效的配置。
}

// ThreadCompactResult 是 thread 上下文压缩操作的结果。
type ThreadCompactResult struct {
	ThreadID     string `json:"threadId"`
	Command      string `json:"command"`
	BeforeTokens int    `json:"beforeTokens"`        // 压缩前 token 数。
	AfterTokens  int    `json:"afterTokens"`         // 压缩后 token 数。
	Compacted    bool   `json:"compacted"`           // 是否实际执行了压缩。
	Estimated    bool   `json:"estimated,omitempty"` // token 数是否为估算值。
}

File: frontend-app/src/shared/api/response-validators/core/thread-config.js
// @ts-check

import { assertBackendResponseObject, assertOnlyResponseKeys, assertResponseRecord, hasOwn, validateStringFields } from '../shared.js';

const THREAD_CONFIG_RESPONSE_KEYS = new Set(['threadId', 'provider', 'supportsThreadOverride', 'override', 'effective']);
const THREAD_CONFIG_VALUES_KEYS = new Set(['model', 'effort', 'approvals']);
const THREAD_COMPACT_RESPONSE_KEYS = new Set(['threadId', 'command', 'beforeTokens', 'afterTokens', 'compacted', 'estimated']);
const THREAD_RECOVER_RESPONSE_KEYS = new Set(['thread', 'recovered', 'mode']);
const THREAD_RECOVER_THREAD_KEYS = new Set(['id', 'status']);

/** @param {string} method @param {unknown} response @param {string} label */
function validateThreadConfigValues(method, response, label) {
  const value = assertResponseRecord(method, response, label);
  assertOnlyResponseKeys(method, value, THREAD_CONFIG_VALUES_KEYS, label);
  validateStringFields(method, value, label, [], ['model', 'effort', 'approvals']);
}

/** @param {string} method @param {unknown} response */
function validateThreadConfigResponse(method, response) {
  const value = assertBackendResponseObject(method, response);
  assertOnlyResponseKeys(method, value, THREAD_CONFIG_RESPONSE_KEYS, 'body');
  validateStringFields(method, value, 'body', ['threadId'], ['provider']);
  if (typeof value.supportsThreadOverride !== 'boolean') {
    throw new TypeError(`${method} response supportsThreadOverride must be a boolean`);
  }
  validateThreadConfigValues(method, value.override, 'override');
  validateThreadConfigValues(method, value.effective, 'effective');
  return value;
}

/** @param {string} method @param {unknown} response */
function validateThreadCompactResponse(method, response) {
  const value = assertBackendResponseObject(method, response);
  assertOnlyResponseKeys(method, value, THREAD_COMPACT_RESPONSE_KEYS, 'body');
  validateStringFields(method, value, 'body', ['threadId', 'command'], []);
  for (const key of ['beforeTokens', 'afterTokens']) {
    if (!Number.isInteger(value[key])) {
      throw new TypeError(`${method} response ${key} must be an integer`);
    }
  }
  if (typeof value.compacted !== 'boolean') {
    throw new TypeError(`${method} response compacted must be a boolean`);
  }
  if (hasOwn(value, 'estimated') && typeof value.estimated !== 'boolean') {
    throw new TypeError(`${method} response estimated must be a boolean`);
  }
  return value;
}

/** @param {string} method @param {unknown} response */
function validateThreadRecoverResponse(method, response) {
  const value = assertBackendResponseObject(method, response);
  assertOnlyResponseKeys(method, value, THREAD_RECOVER_RESPONSE_KEYS, 'body');
  const thread = assertResponseRecord(method, value.thread, 'thread');
  assertOnlyResponseKeys(method, thread, THREAD_RECOVER_THREAD_KEYS, 'thread');
  validateStringFields(method, thread, 'thread', ['id'], ['status']);
  if (typeof value.recovered !== 'boolean') {
    throw new TypeError(`${method} response recovered must be a boolean`);
  }
  if (typeof value.mode !== 'string') {
    throw new TypeError(`${method} response mode must be a string`);
  }
  return value;
}

export { validateThreadCompactResponse, validateThreadConfigResponse, validateThreadRecoverResponse };
